_source/_posts/2021-01-20-reactive-java-microservices.adoc

```diff
@@ -0,0 +1,545 @@
+Install the Okta CLI using the aforementioned website and come back here when you're done. If you don't have an Okta developer account, run `okta register`.
+
+NOTE: You can also use your browser and Okta's developer console to register an app. See https://www.jhipster.tech/security/#okta[JHipster's security documentation] for those instructions.
+
+Then, from the gateway project's directory, run `okta apps create jhipster`. Accept the default redirect URIs.
+
+image::{% asset_path 'blog/reactive-java-microservices/okta-apps-create.png' %}[alt=Okta CLI in action,width=800,align=center]
+
+This process does several things:
+
+. Registers an OIDC app in Okta with JHipster's configured redirect URIs.
+. Creates `ROLE_ADMIN` and `ROLE_USER` groups and adds your user to both.
+. Creates a `groups` claim and adds it to ID tokens.
+. Creates a `.okta.env` file with the values you'll need to talk to Okta.
+
+=== Update the JHipster Registry to Distribute OIDC Configuration
+
+I mentioned earlier that Spring Cloud Config allows you to distribute Spring's configuration between apps. In this section, you'll configure JHipster's Spring Security settings to use Okta across all your services.
+
+Add the following YAML to `gateway/src/main/docker/central-server-config/localhost-config/application.yml`. You can find the values for each property in the `.okta.env` file.
+
+[source,yaml]
+----
+spring:
+  security:
+    oauth2:
+      client:
+        provider:
+          oidc:
+            issuer-uri: https://<your-okta-domain>/oauth2/default
+        registration:
+          oidc:
+            client-id: <client-id>
+            client-secret: <client-secret>
+----
+
+Save your changes. These values will be distributed to the JHipster Registry, gateway, blog, and store apps. Restart the JHipster Registry by running the following commands:
+
+[source,shell]
+----
+docker-compose -f src/main/docker/jhipster-registry.yml down #jhregistrydown
+docker-compose -f src/main/docker/jhipster-registry.yml up -d #jhregistryup
+----
+
+Use kbd:[Ctrl + C] to kill all your `./gradlew` processes and start them again.
+
+Now, open a new incognito browser window, go to `http://localhost:8080`, and sign in. Rejoice that using Okta for authentication works!
+
+image::{% asset_path 'blog/reactive-java-microservices/okta-login.png' %}[alt=Okta login,width=800,align=center]
+
+image::{% asset_path 'blog/reactive-java-microservices/okta-login-success.png' %}[alt=Okta login success,width=800,align=center]
+
+If you're feeling lucky, you can set your Okta credentials as environment variables and run end-to-end tests (from the `gateway` directory).
+
+[source,shell]
+----
+export E2E_USERNAME=<your-username>
+export E2E_PASSWORD=<your-password>
+npm run e2e
+----
+
+CAUTION: Your JHipster Registry is now configured to use Okta for authentication. You'll need to add redirect URIs in your Okta app for `http://localhost:8761`. After making these changes, you should be able to sign in to it.
```

Comment:
Maybe mention that this is test specific configuration. "To make the test work you'll need to add..."

Reply:
It's not really test-specific. It's more "when developing locally" specific. That's why I entered https://github.com/okta/okta-cli/issues/75.

Reply:
AHHH, I totally get it now! That's an easy one